In [1]:
# Otimizar sklearn
from sklearnex import patch_sklearn
patch_sklearn()

Intel(R) Extension for Scikit-learn* enabled (https://github.com/intel/scikit-learn-intelex)


# Imports

In [2]:
import pandas as pd

# Modelos
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier, plot_tree

# Train/Test
from sklearn.model_selection import train_test_split, cross_val_score, cross_val_predict

# Carregar dados

In [3]:
data = pd.read_feather('topic_classifier.feather').sample(frac=1).astype(str)
data

title  \
4695                 any birthday activity suggestions?   
3348  An Important Reminder To The WM Community As W...   
1949    Caught this cool sunrise at Shivaji Park today.   
4087  [P] Output of GAN to match text's input via cl...   
4865  EP 277: Getting Over an Ex and Calling in Love...   
...                                                 ...   
4187               Stack Exchange Podcast - Episode #02   
3597                                          E - F = _   
3156  Moved almost 2 years ago and noticed a ton of ...   
1544               Docker Dev Workflow for Apache Spark   
5693  Different levels of stillness I experience dur...   

                                                 c1body                 Topic  
4695  we're all in our 20s and want do something for...                  News  
3348  As a friendly reminder, our community is inten...         Entertainment  
1949                                  [link] [comments]                  News  
4087  https://preview.redd.it/3o45ur0apdc61.png?widt...      Machine Learning  
4865  This call is about moving out of victim and re...           Sprituality  
...                                                 ...                   ...  
4187  Stack Exchange Podcast - Episode #02 by The St...  Software Development  
3597                                  [link] [comments]         Entertainment  
3156                                  [link] [comments]         Entertainment  
1544  The benefits that come with using Docker conta...      Machine Learning  
5693  I'm pretty new to studying meditation at deepe...           Sprituality  

[5871 rows x 3 columns]

# Rodar modelos

In [4]:
colors = {
    'Entertainment': 'red',
    'Independence': 'orange',
    'Machine Learning': 'green',
    'News': 'blue',
    'Science': 'purple',
    'Software Development': 'brown',
    'Sprituality': 'gray',
    'Wellness': 'cyan'
}


## Logistic Regression

In [5]:
def logistic_regression():
    return LogisticRegression()

### TFIDF

In [6]:
tfidf = pd.read_feather('smote/tfidf.feather').sample(frac=1)
tfidf.head()

00  000  000000  0001  00029890  00072  001  0014  00152  00206  ...  \
1502  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
2051  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
2417  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
4029  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
4508  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   

       ఆయ   గభ   దర   రత   లన   షల  fixed  𝙎𝙝𝙖𝙣𝙠𝙖𝙧𝙖𝙘𝙝𝙖𝙧𝙮𝙖  𝙏𝙚𝙢𝙥𝙡𝙚  \
1502  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0   
2051  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0   
2417  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0   
4029  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0   
4508  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0   

                     Topic  
1502      Machine Learning  
2051  Software Development  
2417           Sprituality  
4029         Entertainment  
4508               Science  

[5 rows x 32327 columns]

In [7]:
y_true = tfidf['Topic']
X = tfidf.drop(columns=['Topic'])

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y_true, test_size=0.3)
print(f'{len(X_train)}')
print(f'{len(y_train)}')
print(f'{len(X_test)}')
print(f'{len(y_test)}')

3796
3796
1628
1628


#### Separar seção de treino/teste

In [9]:
model = logistic_regression()
model.fit(X_train, y_train)
score = model.score(X_test, y_test)
print("Score para todos")
score

Score para todos


0.894963144963145

In [21]:
for topic in colors:
    # score = model.score(X_test, y_test)
    y_true = tfidf['Topic'].apply(lambda x: topic if x == topic else f'Não {topic}')
    X = tfidf.drop(columns=['Topic'])
    print(f'Score para {topic}')
    print(X.head(), y_true.head())
    continue
    X_train, X_test, y_train, y_test = train_test_split(X, y_true)
    model = logistic_regression()
    model.fit(X_train, y_train)
    print(model.score(X_test, y_test))

Score para Entertainment
       00  000  000000  0001  00029890  00072  001  0014  00152  00206  ...  \
1502  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
2051  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
2417  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
4029  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   
4508  0.0  0.0     0.0   0.0       0.0    0.0  0.0   0.0    0.0    0.0  ...   

       வர   ఆయ   గభ   దర   రత   లన   షల  fixed  𝙎𝙝𝙖𝙣𝙠𝙖𝙧𝙖𝙘𝙝𝙖𝙧𝙮𝙖  𝙏𝙚𝙢𝙥𝙡𝙚  
1502  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0  
2051  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0  
2417  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0  
4029  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0  
4508  0.0  0.0  0.0  0.0  0.0  0.0  0.0   0.0             0.0     0.0  

[5 rows x 32326 columns] 1502    Não Entertainment
2051    Não Ente

Score para todos
0.894963144963145

Score para Entertainment
0.915929203539823

Score para Independence
0.9299410029498525

Score para Machine Learning
0.9365781710914455

Score para News
0.9033923303834809

Score para Science
0.9218289085545722

Score para Software Development
0.943952802359882

Score para Sprituality
0.9247787610619469

Score para Wellness
0.9837758112094396

In [22]:
y_pred = model.predict(X_test)
y_pred
print(f'{y_pred=} {y_true=}')

y_pred=array(['Não Wellness', 'Wellness', 'Wellness', ..., 'Wellness',
       'Não Wellness', 'Wellness'], dtype=object) y_true=1502    Não Wellness
2051    Não Wellness
2417    Não Wellness
4029    Não Wellness
4508    Não Wellness
            ...     
1124    Não Wellness
4148    Não Wellness
4340    Não Wellness
4375    Não Wellness
1810    Não Wellness
Name: Topic, Length: 5424, dtype: object


#### Cross Validation

In [15]:
scores = cross_val_score(logistic_regression(), X, y_true, cv=10, verbose=1, n_jobs=3)
print(scores, scores.mean())

[Parallel(n_jobs=3)]: Using backend LokyBackend with 3 concurrent workers.
[Parallel(n_jobs=3)]: Done  10 out of  10 | elapsed:  3.5min finished


[0.98710866 0.99447514 0.98710866 0.98895028 0.98708487 0.99077491
 0.99261993 0.99077491 0.9797048  0.99077491] 0.9889377042941699


In [18]:
for topic in colors:
    # score = model.score(X_test, y_test)
    y_true = tfidf['Topic'].apply(lambda x: topic if x == topic else f'Não {topic}')
    X = tfidf.drop(columns=['Topic'])
    print(f'Score para {topic}')
    scores = cross_val_score(logistic_regression(), X, y_true, cv=10, verbose=1, n_jobs=2)
    print(scores, scores.mean())
    print()

Score para Entertainment


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  3.2min finished


[0.92081031 0.93186004 0.92081031 0.9281768  0.92066421 0.91512915
 0.91328413 0.91512915 0.91143911 0.91512915] 0.9192432366312613

Score para Independence


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  3.1min finished


[0.95764273 0.97053407 0.946593   0.9373849  0.9501845  0.9501845
 0.94464945 0.94464945 0.95940959 0.92804428] 0.9489276467350309

Score para Machine Learning


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  3.0min finished


[0.9373849  0.94843462 0.95211786 0.9281768  0.94095941 0.94464945
 0.93726937 0.94649446 0.93542435 0.93911439] 0.9410025619593213

Score para News


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  3.0min finished


[0.91344383 0.91344383 0.91344383 0.91528545 0.91881919 0.91328413
 0.90590406 0.91143911 0.9095941  0.91512915] 0.9129786684607177

Score para Science


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  2.8min finished


[0.92633517 0.91712707 0.92449355 0.91712707 0.9298893  0.92435424
 0.92619926 0.91881919 0.92804428 0.93173432] 0.9244123463334081

Score para Software Development


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  3.0min finished


[0.96316759 0.95395948 0.96869245 0.9558011  0.93911439 0.96309963
 0.94833948 0.95571956 0.96125461 0.94833948] 0.95574877848226

Score para Sprituality


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  2.8min finished


[0.92633517 0.92633517 0.94106814 0.93186004 0.93726937 0.94649446
 0.93173432 0.94280443 0.93173432 0.93542435] 0.9351059781316045

Score para Wellness


[Parallel(n_jobs=2)]: Using backend LokyBackend with 2 concurrent workers.
[Parallel(n_jobs=2)]: Done  10 out of  10 | elapsed:  3.0min finished


[0.98710866 0.99447514 0.98710866 0.98895028 0.98708487 0.99077491
 0.99261993 0.99077491 0.9797048  0.99077491] 0.9889377042941699



Scores para todos
[0.98710866 0.99447514 0.98710866 0.98895028 0.98708487 0.99077491
 0.99261993 0.99077491 0.9797048  0.99077491] 0.9889377042941699

Score para Entertainment
[0.92081031 0.93186004 0.92081031 0.9281768  0.92066421 0.91512915
 0.91328413 0.91512915 0.91143911 0.91512915] 0.9192432366312613

Score para Independence
[0.95764273 0.97053407 0.946593   0.9373849  0.9501845  0.9501845
 0.94464945 0.94464945 0.95940959 0.92804428] 0.9489276467350309

Score para Machine Learning
[0.9373849  0.94843462 0.95211786 0.9281768  0.94095941 0.94464945
 0.93726937 0.94649446 0.93542435 0.93911439] 0.9410025619593213

Score para News
[0.91344383 0.91344383 0.91344383 0.91528545 0.91881919 0.91328413
 0.90590406 0.91143911 0.9095941  0.91512915] 0.9129786684607177

Score para Science
[0.92633517 0.91712707 0.92449355 0.91712707 0.9298893  0.92435424
 0.92619926 0.91881919 0.92804428 0.93173432] 0.9244123463334081

Score para Software Development
[0.96316759 0.95395948 0.96869245 0.9558011  0.93911439 0.96309963
 0.94833948 0.95571956 0.96125461 0.94833948] 0.95574877848226

Score para Sprituality
[0.92633517 0.92633517 0.94106814 0.93186004 0.93726937 0.94649446
 0.93173432 0.94280443 0.93173432 0.93542435] 0.9351059781316045

Score para Wellness
[0.98710866 0.99447514 0.98710866 0.98895028 0.98708487 0.99077491
 0.99261993 0.99077491 0.9797048  0.99077491] 0.9889377042941699

### BOW
